# Using the OPNMF method to dimensionally reduce the ADNI imaging data

In [1]:
import pandas as pd
import numpy as np
import matplotlib

### Read the database, examine it:

Instead of reading the whole database, we read only the data that's useful to us. That is, we read only specific columns of data, and we take only the row containing the first scan for each person. 

In "ADNI Regressional Analysis.ipynb" we have done that exactly, as well as performed linear regression transformation to the imaging data, in order to remove any age, sex, and DLICV_baseline effect. 

The data is located at "./DATA/Reduced_Linearly_Transformed_Unique_Dataset.pkl"

In [2]:
unique = pd.read_pickle("./DATA/Linearly_Transformed_Unique_Dataset.pkl")
unique.head(15)

PTID        Date        Age  Sex  DLICV_baseline  APOE4_Alleles  \
0    002_S_0295  2006-04-18  84.742466    0     1485405.375            1.0   
9    002_S_0413  2006-05-02  76.283562    1     1364116.000            0.0   
24   002_S_0559  2006-05-23  79.223288    0     1570479.625            1.0   
31   002_S_0619  2006-06-01  77.447945    0     1859348.250            2.0   
36   002_S_0685  2006-07-06  89.561644    1     1372862.125            0.0   
45   002_S_0729  2006-07-17  65.056164    1     1166961.750            1.0   
64   002_S_0816  2006-08-30  70.767123    0     1444128.125            2.0   
69   002_S_0938  2006-10-05  82.167123    1     1309685.000            0.0   
74   002_S_0954  2006-10-10  69.198630    1     1075661.500            1.0   
81   002_S_0955  2006-10-11  78.161644    1     1363607.000            1.0   
84   002_S_1018  2006-11-29  70.658904    1     1355603.000            0.0   
90   002_S_1070  2006-11-28  73.564384    0     1550701.375            0.0   
97   002_S_1155  2006-12-14  57.742466    0     1472727.750            0.0   
113  002_S_1261  2007-02-15  71.067123    1     1350714.875            0.0   
127  002_S_1268  2007-02-14  82.642466    0     1435189.875            1.0   

    APOE_Genotype Diagnosis_nearest_2.0  MUSE_Volume_4  MUSE_Volume_11  ...  \
0           E3/E4                    CN    -440.777069     -507.297168  ...   
9           E3/E3                    CN     577.755137     -188.813792  ...   
24          E3/E4                    CN     198.499249     1080.290951  ...   
31          E4/E4              Dementia    2623.687012      649.558822  ...   
36          E3/E3                    CN     727.400228     -250.782773  ...   
45          E3/E4                   MCI     246.226215      628.340793  ...   
64          E4/E4              Dementia    -145.138564     -193.593195  ...   
69          E3/E3              Dementia     169.421397     -610.085153  ...   
74          E3/E4                   MCI     -81.664210     1343.833768  ...   
81          E3/E4              Dementia    1036.385233     -353.324662  ...   
84          E3/E3              Dementia    -495.018850      486.447691  ...   
90          E3/E3                   MCI     248.001697      515.447841  ...   
97          E3/E3                   MCI     129.951985      -16.174703  ...   
113         E3/E3                    CN    -213.577546      489.130696  ...   
127         E3/E4                   MCI     208.117466     -496.119467  ...   

     rs28394864  rs111278892  rs3752246  rs4147929  rs41289512  rs3865444  \
0             0            1          1          1           0          0   
9             1            0          1          1           0          1   
24            2            0          0          0           0          1   
31            1            0          0          0           1          1   
36            2            1          1          1           0          0   
45            1            0          0          0           1          1   
64            0            0          0          0           0          1   
69            0            0          1          1           0          1   
74            2            2          1          1           0          1   
81            1            1          0          0           0          1   
84            0            1          1          1           0          0   
90            2            0          0          0           0          0   
97            1            1          0          0           0          0   
113           1            0          0          0           0          0   
127           1            0          0          0           0          1   

     rs6024870  rs6014724  rs7274581  rs429358  
0            0          0          0         1  
9            0          0          0         0  
24           0          0          0         0  
31           0          0          0         2  
36           0     

### The columns that interest us are the 145 ROIs:

We need essentially the columns MUSE_Volume_4 , ... , MUSE_Volume_207. These contain the residuals from the Linear Regression performed previously. 

In [6]:
unique.loc[:,"MUSE_Volume_4":"MUSE_Volume_207"].head(15)

MUSE_Volume_4  MUSE_Volume_11  MUSE_Volume_23  MUSE_Volume_30  \
0      -440.777069     -507.297168      -35.171797      -14.510211   
9       577.755137     -188.813792       35.574764      -39.881572   
24      198.499249     1080.290951      137.416288      142.586830   
31     2623.687012      649.558822     -162.939446     -122.191780   
36      727.400228     -250.782773      -27.195126        6.992014   
45      246.226215      628.340793       10.979183       24.346908   
64     -145.138564     -193.593195       -1.961382      -43.963703   
69      169.421397     -610.085153      -82.243440      -88.850034   
74      -81.664210     1343.833768      101.836384       43.596098   
81     1036.385233     -353.324662      -49.617421      -18.439395   
84     -495.018850      486.447691      -65.242657      -47.217117   
90      248.001697      515.447841       -4.335870      -51.341650   
97      129.951985      -16.174703       70.933092       61.656025   
113    -213.577546      489.130696      -52.240459      -32.707803   
127     208.117466     -496.119467      -25.244931      -65.393865   

     MUSE_Volume_31  MUSE_Volume_32  MUSE_Volume_35  MUSE_Volume_36  \
0         90.244138      140.075639      299.828133       63.889680   
9         40.161648       58.255314     -909.956510     -107.325098   
24       121.231074       41.449232     1825.886437     -267.694901   
31      -329.934406     -351.510297    -3426.992838     -826.297201   
36        48.594085       57.675286     1120.707525      136.378936   
45      -165.999584     -114.587813      171.517739      628.498317   
64      -114.956569     -287.346946      993.225831     -355.357204   
69      -388.425224     -393.192570     -890.491741     -746.264217   
74      -154.810341     -134.886347     1033.170275      968.539022   
81      -164.007463     -327.518454      717.249036      257.486180   
84      -136.476165     -118.658886     2624.773739     -481.177039   
90         3.707762      -57.156866     1743.878904     -347.518465   
97       -41.714673      -85.693524    -1123.797096       51.767167   
113     -194.553503     -214.660995     -893.886856      -18.736473   
127      -49.082556       28.265007     1251.674629      109.992557   

     MUSE_Volume_37  MUSE_Volume_38  ...  MUSE_Volume_198  MUSE_Volume_199  \
0         56.259492     4434.963481  ...       745.557312      -188.954470   
9        118.445748     -932.590538  ...      1336.384182      2631.004114   
24         6.605333     -947.176391  ...     -1764.158370     -2206.292278   
31      -713.213854     -355.750507  ...      -641.454806       583.322773   
36       316.063586     4364.772160  ...        47.451381       279.144879   
45        88.705570     -521.590388  ...      -346.626209      -670.579403   
64      -299.109526    -4585.096939  ...      -639.210338      1881.904887   
69      -621.836515    -5005.238233  ...      -891.150119       190.989398   
74       941.960498    -1405.349490  ...     -2610.451083      -841.474614   
81       312.588260      351.860893  ...     -2120.599430       -62.905455   
84      -339.522304     5135.070563  ...       524.431513      -673.056464   
90      -245.438625     4732.917243  ...      -421.858244      -807.272666   
97       343.965742    -4021.276684  ...     -1130.593896      -155.368851   
113      -21.610104     2650.325801  ...       475.602113       327.233504   
127      236.986827     3695.547106  ...        99.233224      2383.736063   

     MUSE_Volume_200  MUSE_Volume_201  MUSE_Volume_202  MUSE_Volume_203  \
0       -1594.432454     -1648.308374       798.003198      -468.672456   
9        1410.754665        30.295558     -1258.071206       115.187177   
24       1473.087979       532.054466      1714.763199      2469.640085   
31       -701.560285     -1369.412583     -2919.253412     -2766.270514   
36      -1423.470776     -1844.920979      -186.886077      -270.949818   
45        163.045892      1008.186971     -1557.